### PR TITLE
feat: join error

### LINKED ISSUE #22: feat: support Join
Similar to `errors.Join`: https://pkg.go.dev/errors#Join

`eris.Is` will return true for all Joined errors so that we can have polymorphism for an error.

Reference:
https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/errors/join.go;l=36
https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/errors/wrap.go;l=63

### PR DESCRIPTION
**Pull request by Gogomoe**
## What's changed and what's your intention?
resolve #22 
`eris.Join(err...)` almost is `eris.Wrap(errors.Join(err...), "join error")`
Implementing and testing the format is a bit hard.
the format looks like:
```
// no trace text
code(internal) outer wrap: code(unknown) join error
0>	fmt error
1>	code(internal) wrap1: code(internal) wrap2: external
2>	code(unknown) eris error
```
```
// trace text
code(internal) outer wrap
	eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:22
code(unknown) join error
	eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:22
	eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:17
0>	fmt error
1>	code(internal) wrap1
		eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:19
	code(internal) wrap2
		eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:19
		eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:19
	external
2>	code(unknown) eris error
		eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:20
```

```
// no trace json
{
	"externals": [
		{
			"external": "fmt error"
		},
		{
			"external": "external",
			"root": {
				"code": "internal",
				"message": "wrap2"
			},
			"wrap": [
				{
					"code": "internal",
					"message": "wrap1"
				}
			]
		},
		{
			"root": {
				"code": "unknown",
				"message": "eris error"
			}
		}
	],
	"root": {
		"code": "unknown",
		"message": "join error"
	},
	"wrap": [
		{
			"code": "internal",
			"message": "outer wrap"
		}
	]
}
```
```
// trace json
{
	"externals": [
		{
			"external": "fmt error"
		},
		{
			"external": "external",
			"root": {
				"code": "internal",
				"message": "wrap2",
				"stack": [
					"eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:19",
					"eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:19"
				]
			},
			"wrap": [
				{
					"code": "internal",
					"message": "wrap1",
					"stack": "eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:19"
				}
			]
		},
		{
			"root": {
				"code": "unknown",
				"message": "eris error",
				"stack": [
					"eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:20"
				]
			}
		}
	],
	"root": {
		"code": "unknown",
		"message": "join error",
		"stack": [
			"eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:22",
			"eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:17"
		]
	},
	"wrap": [
		{
			"code": "internal",
			"message": "outer wrap",
			"stack": "eris_test.TestName:/Users/gogo/workspace/eris/a_test.go:22"
		}
	]
}
```
## Checklist

- [ ] I have written the necessary docs and comments
- [ ] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
Close #22

### CODE REVIEW

File: eris.go [hunk 1]
```diff
@@ -95,6 +96,19 @@ func Errorf(format string, args ...any) statusError {
 	}
 }
 
+type joinError interface {
+	Unwrap() []error
+}
+
+// Join returns an error that wraps the given errors.
+func Join(errs ...error) error {
+	internal := errors.Join(errs...)
```
Comment (wjf3121) [inline eris.go:105]: This will break `eris.Is`'s behavior, which I'm not so sure.

Caller can be easily mistaken and make the wrong assumption.

Comment (CAJan93) [reply eris.go:105]: I think we should for sure test that `eris.Is` still works with this implementation 

First simple test in https://github.com/risingwavelabs/eris/pull/28 shows that it still seems to work. In the end we inherit the behaviour of [errors.Join](https://pkg.go.dev/errors#Join)

I did not want to push into your branch, that is why I created the tiny PR @Gogomoe

Comment (Gogomoe) [reply eris.go:105]: Sorry i forget to reply the comment. The eris.is and as are implemented in #27. I also added more testcases to cover the join error

**Comment by CAJan93** [comment]
@Gogomoe out of curiosity. What would you like to use `eris.Join` for? To me the idiomatic way to handle go errors is 

```go 
value, err := someFunc(input)
if err != nil {
   // wrap err and return
}
```

When do we want to join? The only things that I can think about are
- validating something, e.g. user input, and returning an all the found errors 
- doing multiple tasks concurrently, waiting until all tasks are done, and then joining the errors.

**Comment by Gogomoe** [comment]
> @Gogomoe out of curiosity. What would you like to use `eris.Join` for? To me the idiomatic way to handle go errors is
> 

Another situation is retry an action for mutiple times. It's a common pattern used in our code base. I dont' want to ignore any error when all retry fails. Currently, we use errors.join in this case. You can check the use case by searching it in our code base

File: format_test.go [hunk 2]
```diff
@@ -393,3 +395,190 @@ func TestFormatJSONWithStack(t *testing.T) {
...
+				}
+				for i := 0; i < len(wrapMap); i++ {
+					if _, exists := wrapMap[i]["message"]; !exists {
+						t.Fatalf("%v: expected a 'message' field in the output but didn't find one { %v }", desc, errJSON)
+					}
+					if wrapMap[i]["message"] != tt.wrapOutput[i]["message"] {
+						t.Errorf("%v: expected { %v } got { %v }", desc, tt.wrapOutput[i]["message"], wrapMap[i]["message"])
+					}
+					if _, exists := wrapMap[i]["stack"]; tt.withTrace && !exists {
+						t.Fatalf("%v: expected a 'stack' field in the output but didn't find one { %v }", desc, errJSON)
```
Comment (wjf3121) [inline format_test.go:550]: nit: I suppose this can be `t.Errorf` as we can still check the rest of the wrapError in the loop.

**Review by wjf3121 — APPROVED** [review]
LGTM!

File: format_test.go (hunk 2 above)
Comment (Gogomoe) [reply format_test.go:550]: I copied the `wrapMap` section from other test cases. Leet's keep it 🤪

**Comment by Gogomoe** [comment]
I plan to merge this PR tomorrow. @CAJan93  Do you have more opinions about this PR?

**Pull request merged**